### PR TITLE
Ported the vrf verification feature

### PR DESCRIPTION
**Pull request by adon-at-work**
This is the vrf verification implementation ported from https://github.com/yahoo/coname/blob/master/vrf/vrf.go 

It lives as an utility function, and is tested with inputs and outputs captured from the original golang implementation.

Please review, thanks.

### CODE REVIEW

File: src/javascript/crypto/e2e/vrf/sha3shake256.js [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+goog.require('goog.asserts');
+
+/*
+ * Imported from js-sha3 v0.3.1
+ * https://github.com/emn178/js-sha3
+ *
+ * Copyright 2015, emn178@gmail.com
+ * 
+ * Licensed under the MIT license:
+ * http://www.opensource.org/licenses/MIT
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: Should this go under the `@license` header? I'm not sure what is the recommended way to include the MIT license in an Apache2 project. @gyehuda

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+goog.provide('e2e.vrf.extra25519');
+
+goog.require('goog.asserts');
+goog.require('goog.crypt');
+goog.require('e2e.BigNum');
+goog.require('e2e.ecc.constant');
+goog.require('e2e.ecc.DomainParam');
+goog.require('e2e.ecc.PrimeCurve');
+
+goog.require('e2e.vrf.sha3Shake256');
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Nit: requires should be in alphabetical order. You can run `./do.sh lint` on the newly-added files to find linting errors.

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+ * Refer to Definition 1 in https://www.iacr.org/archive/pkc2003/25670211/25670211.pdf
+ * 
+ * @param {!e2e.ByteArray} sBytes A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.extra25519.fromBytesBaseGroup = function(sBytes) {
+
+	var P = e2e.vrf.ed25519.curve.pointFromByteArray(sBytes);
+
+	goog.asserts.assert(!P.isInfinity(),
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Are assertions the right tool here? I am used to using assertions to detect programmer errors, but a completely good user running perfect code could still be sent an invalid point by a malicious server.

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	// goog.asserts.assert(!P.isIdentity(),
+        'P != infinity');
+
+	goog.asserts.assert(P.isOnCurve(),
+        'P must satisfy the defining equation of the elliptic curve E');
+
+	// TODO: confirm https://github.com/google/end-to-end/issues/340
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: I have no idea what is going on there. Multiplication by order really should give the identity element. This check is important, though -- we should not launch with it simply omitted.

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: I sort of follow the practise used by the ecc library. [Example 1](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/point/curve25519.js#L71-L72) [Example 2](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/curve/curve25519.js#L85)

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 4 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Right now, `P.multiply(e2e.vrf.ed25519.n).isInfinity()` evaluates to `false`, whereas `P.multiply(e2e.vrf.ed25519.n).isIdentity()` evaluates to `true`.

The underlying function calls can be found at here: [isInfinity(), isIdentity()](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/point/ed25519.js#L162-L174), and the points [INFINITY, IDENTITY](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/curve/ed25519.js#L103-L117).

I was curious but didn't have time to figure out yet why the infinity point is different than the identity point under the ed25519 implementation, while they're equal under the [curve25519 one](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/point/curve25519.js#L104-L118).

The check at the [golang implementation](https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1117) is apparently checking against the identity point. So I wondered if `isIdentity()` is what we want to check at here (line 58). Same for line 48? Any ideas?

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 2 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: thanks. will fix today when I get back to office. i dont have that lint tool in my personal laptop. :)

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 4 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: @andres-erbsen, I'm so glad we can have your reviews here. You're so awesome. thanks. :+1:

File: src/javascript/crypto/e2e/vrf/sha3shake256.js (hunk 1 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: gd point. i was curious what to do too. I copied the code from there, dropped some, and added some new features. so, what's the license then? follow the MIT? or stick with the Apache as used in other files? or dual licenses? trust the expert @gyehuda can give some insights. thanks

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: I just had another look at https://github.com/yahoo/coname/blob/master/vrf/vrf.go#L171 and https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1095-L1127, even a check failed and returned `false`, [p is already filled](https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1103) and your code is apparently ignoring what is returned. Is that expected?

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 2 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: the changes are prepared at https://gist.github.com/adon-at-work/ca254eed709ecf322d47
will commit it to this PR at last, as to avoid collapsing all the comments due to "outdated diff".

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 4 above)
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Hmm. I think isIdentity is the way to go, and I am not sure what meaning `this.INFINITY = new e2e.ecc.point.Ed25519(this,  this.ZERO, this.ZERO, this.ZERO, this.ZERO)` has.

The reason we want this check is that we want all points used in the vrf computation to be in the prime-order subgroup of the non-prime-order elliptic curve. Multiplying by order of the subgroup should give the identity element (of both the subgroup and the curve) in that case. This reasoning does not involve the geometric ("at infinity") interpretation at all.

http://iml.univ-mrs.fr/ati/crypto_puces/2011/slides/Mardi/rolland.pdf and https://christianepeters.files.wordpress.com/2012/10/20080620-rennes.pdf say that projective(0,1,0) (the identity) is a point at infinity. https://hyperelliptic.org/EFD/oldefd/edwards.html and https://en.wikipedia.org/wiki/Edwards_curve#Projective_homogeneous_coordinates confirm that that is the neutral element (with no mention of infinity).

Traditional NIST curves only have one point at infinity and this is the identity. In edwards curves, we cannot assume that. Which still does not explain the value of `INFINITY` this library defines for ed25519, but oh well. I also find it interesting that the ed25519 code comments mention jacobian coordinates but reference the EFD entry for addition law in the (extended) projective coordinates.

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: That is the code I wrote and intended to write. In retrospect it would have been nicer to actually zero the output when an error is detected (but no harm done since the returned value does not get used).

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+	var e = v3.add(v2.multiply(A)).add(v)
+			.power(e2e.vrf.ed25519.curve.q.subtract(e2e.BigNum.ONE).shiftRight(1));
+
+	// leading zeros are dropped
+	// assertEquals('01',
+	// 	goog.crypt.byteArrayToHex(e.x.toByteArray().reverse()));
+
+	// e is either 1 or -1
+	// x = ev - (1-e)A/2
+	var x = e.isEqual(e2e.vrf.ed25519.curve.ONE) ? v : A.substract(v);
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Please document that this is not constant time and thus must not be used with secret inputs.

File: src/javascript/crypto/e2e/vrf/sha3shake256.js [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    0, 2147516545, 2147483648, 32777, 2147483648, 138, 0, 136, 0, 2147516425, 0,
+    2147483658, 0, 2147516555, 0, 139, 2147483648, 32905, 2147483648, 32771,
+    2147483648, 32770, 2147483648, 128, 2147483648, 32778, 0, 2147483658, 2147483648,
+    2147516545, 2147483648, 32896, 2147483648, 2147483649, 0, 2147516424, 2147483648
+];
+
+var blocks = [],
+    s = [];
+
+var keccak = function(message, bits, padding, outputByteLen, outputFormat) {
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: I am not sure what the constant-time status of this is. "yes" for non-strings and "no" for strings? Maybe just err on the side of safety and say it is not necessarily constant time.

File: src/javascript/crypto/e2e/vrf/vrf_test.html [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+/*
+Test Vector 1
+pk:           C2BF608B6AF918FDC68358A964E707D346ABABCF181E96724D7CF07BBF0E1D6F
+sk:           4AFC605963BA3572D8547B371CE828C822B4B26B21F52838DB33BE7833583321C2BF608B6AF918FDC68358A964E707D346ABABCF181E96724D7CF07BBF0E1D6F
+alice(bytes): 616C696365
+aliceVRF:     4462374E99BD0B0F08EE8405351CE816DE62158B59436FC5D54B56E2B247F59F
+aliceProof:   318004FD67F1A433152DA837126716E9F588F2EE71DAA07A81594867FADE0303EFEB5D62AD73A818DEA5BAE08A4CC9ED82C92F12BF183D50718BF6E9175EB10CC16A26AC42C016BCB10E906426B36046379D508B913E5EC3B5E0B72A404291A2
+*/
+
+var pk = [194, 191, 96, 139, 106, 249, 24, 253, 198, 131, 88, 169, 100, 231, 7, 211, 70, 171, 171, 207, 24, 30, 150, 114, 77, 124, 240, 123, 191, 14, 29, 111];
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/vrf_test.html]: It would be great if the same test vector was in the Go library as well and they linked to each other in comments.

File: src/javascript/crypto/e2e/vrf/vrf_test.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+var alice = [97, 108, 105, 99, 101];
+var aliceVRF = [68, 98, 55, 78, 153, 189, 11, 15, 8, 238, 132, 5, 53, 28, 232, 22, 222, 98, 21, 139, 89, 67, 111, 197, 213, 75, 86, 226, 178, 71, 245, 159];
+var aliceProof = [49, 128, 4, 253, 103, 241, 164, 51, 21, 45, 168, 55, 18, 103, 22, 233, 245, 136, 242, 238, 113, 218, 160, 122, 129, 89, 72, 103, 250, 222, 3, 3, 239, 235, 93, 98, 173, 115, 168, 24, 222, 165, 186, 224, 138, 76, 201, 237, 130, 201, 47, 18, 191, 24, 61, 80, 113, 139, 246, 233, 23, 94, 177, 12, 193, 106, 38, 172, 66, 192, 22, 188, 177, 14, 144, 100, 38, 179, 96, 70, 55, 157, 80, 139, 145, 62, 94, 195, 181, 224, 183, 42, 64, 66, 145, 162];
+
+/**
+ * Tests that e2e.vrf.verify() works
+ */
+function testVRFVerify() {
+    assertTrue(e2e.vrf.verify(pk, alice, aliceVRF, aliceProof));
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/vrf_test.html]: Please add a negative test.

File: src/javascript/crypto/e2e/vrf/sha3shake256.js (hunk 6 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: we can just comment out the string implementation

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+	// edwards25519.GeDoubleScalarMultVartime(&A, &c, &P, &t)
+	// A.ToBytes(&ABytes)
+	var ABytes = e2e.vrf.extra25519.geDoubleScalarMultVartime(c, P, t).toByteArray();
+
+	// hm := hashToCurve(m)
+	// edwards25519.GeDoubleScalarMultVartime(&hmtP, &t, hm, &[32]byte{})
+	// edwards25519.GeDoubleScalarMultVartime(&iicP, &c, &ii, &[32]byte{})
+	var hm = hashToCurve(m);
+	var hmtP = e2e.vrf.extra25519.geDoubleScalarMultVartime(t, hm, e2e.BigNum.ZERO);
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/vrf.js]: The Go code called `geDoubleScalarMultVartime` because it was faster than the single multiplication. This is probably not true here. Which does not necessarily mean that we want to change this code, but if ever performance becomes an issue, this is a place to look.

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+	// edwards25519.ScReduce(&cRef, &cH)
+	var cRef = ed25519.n.residue(new e2e.BigNum(hash.reverse()));
+
+	// return cRef == c
+	return cRef.isEqual(c);
+}
+
+
+/**
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/vrf.js:172]: This documentation does not seem to match the function it is next to.

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: My observation is that: assume the following `pk` used in `FromBytesBaseGroup()` fail to satisfy `nP == IDENTITY`. 
The calculation will continue to produce the subsequent `ABytes`, and so forth...

``` go
P.FromBytesBaseGroup(&pk)
edwards25519.GeDoubleScalarMultVartime(&A, &c, &P, &t)
A.ToBytes(&ABytes)
```

So, indeed the check on `nP == IDENTITY` (and the `P.toBytes(P.fromBytes(pk)) == pk`) is actually optional. 
Therefore, do you think I should just remove all the checks below as to conform to your implementation?

In fact, something like follows makes more sense to me:

``` go
if !P.FromBytesBaseGroup(&pk) {
    return false
}
```

Thoughts?

File: src/javascript/crypto/e2e/vrf/vrf.js (hunk 10 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf.js:172]: gd catch. I noticed that and have planned to fix it as in:
https://gist.github.com/adon-at-work/ca254eed709ecf322d47#file-fixed-gjshint-errors-diff-L1101-L1115

File: src/javascript/crypto/e2e/vrf/vrf.js (hunk 9 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf.js]: Agree. You have a better name for this function? or should I just remove this function call.

File: src/javascript/crypto/e2e/vrf/vrf_test.html (hunk 8 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf_test.html]: agree. i actually tried to randomly alter a byte by hand, and it returned false. :+1:

File: src/javascript/crypto/e2e/vrf/extra25519_test.html [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+var pk = [194, 191, 96, 139, 106, 249, 24, 253, 198, 131, 88, 169, 100, 231, 7, 211, 70, 171, 171, 207, 24, 30, 150, 114, 77, 124, 240, 123, 191, 14, 29, 111];
+var alice = [97, 108, 105, 99, 101];
+// var aliceVRF = [68, 98, 55, 78, 153, 189, 11, 15, 8, 238, 132, 5, 53, 28, 232, 22, 222, 98, 21, 139, 89, 67, 111, 197, 213, 75, 86, 226, 178, 71, 245, 159];
+var aliceProof = [49, 128, 4, 253, 103, 241, 164, 51, 21, 45, 168, 55, 18, 103, 22, 233, 245, 136, 242, 238, 113, 218, 160, 122, 129, 89, 72, 103, 250, 222, 3, 3, 239, 235, 93, 98, 173, 115, 168, 24, 222, 165, 186, 224, 138, 76, 201, 237, 130, 201, 47, 18, 191, 24, 61, 80, 113, 139, 246, 233, 23, 94, 177, 12, 193, 106, 38, 172, 66, 192, 22, 188, 177, 14, 144, 100, 38, 179, 96, 70, 55, 157, 80, 139, 145, 62, 94, 195, 181, 224, 183, 42, 64, 66, 145, 162];
+
+/**
+ * Tests that e2e.vrf.extra25519.fromBytesBaseGroup works, and no assertion error is thrown
+ */
+function testFromBytesBaseGroup() {
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519_test.html:36]: We actually want better tests for this; even just code coverage is better than nothing here. My code does not seem to have any either, sorry about that. I do not have time to implement this right now and I am not sure when I will. So I am just writing something here with no well-thought intent. Some things that are probably true, and if they are, would be useful to check:
- Any multiple of the base point should unmarshal correctly.
- The blacklisted points at http://cr.yp.to/ecdh.html#validate should return error (but the list there is in X25519 format, not ed25519)
- The identity point should return error.
- The points of order 2 and of order 4 from https://christianepeters.files.wordpress.com/2012/10/20080710-s3cm.pdf should return error.
- x_B + y_S should return error if B is the base point and S is a point of order 2 or order 4 and y is less than its order.

There is an additional complication that a normal elliptic curve library might not even care to support the "bad" points, so using them for testing might end up being awkward. In particular, multiplication y*S does clear the low bits of y in some libraries, which would not work here.

File: src/javascript/crypto/e2e/vrf/vrf_test.html (hunk 7 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf_test.html]: Do you expect something like the following: 
https://github.com/adon-at-work/coname/blob/debug/bytedump/vrf/vrf_test.go

I can submit a PR to add such a test case for the golang implementation

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 2 above)
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Ok great. I will run tests and do a build after that pull request.

**Comment by adon-at-work** [comment]
@andres-erbsen , once you finished reviewing this. I'll clear all the commented golang here. They're there to facilitate reviews only. OK?

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 5 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: ok

**Comment by adon-at-work** [comment]
@andres-erbsen and @diracdeltas please check those things fixed.

Since many of those discussions are collapsed since the last commit, I list out the following discussions that are yet to be resolved for quick reference (kindly click below to have them expanded, or you may like to check each collapsed one by one):
- https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39409648
- https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39459701
- https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39517955

File: src/javascript/crypto/e2e/vrf/extra25519_test.html (hunk 11 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519_test.html:36]: I prefer to leave this as a separate PR. OK?

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: note: I changed the implementation from using assertion to throw

but please confirm whether the checks should be enforced in a way that when they failed, we fail the whole vrf verification by returning false.

File: src/javascript/crypto/e2e/vrf/extra25519_test.html (hunk 11 above)
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519_test.html:36]: Yep, that was the intent.

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: The checks are not optional, and are actually very important. My Go implementation was wrong. https://github.com/yahoo/coname/pull/26 is a thing that should happen (and soon after this PR it should also get tests).

File: src/javascript/crypto/e2e/vrf/vrf.js (hunk 9 above)
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/vrf.js]: I think `Vartime` means "not necessarily constant time", so the name is fine. I have no opinion on removing (by inlining and then simplifying).

File: src/javascript/crypto/e2e/vrf/sha3shake256.js (hunk 1 above)
Comment (gyehuda) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: Sorry for the delay. There are a couple of ways to address this, and it's mostly driven by project convention. Some projects have a strict process where the headers are kept consistent and that all inclusions like this are disclosed in a NOTICES file, other projects are much less strict and have what you see above -- an inclusion in the code that is preceded with the relevant license and copyright text. Both are OK. At first this latter approach is simpler, eventually the code could get messy and you'll want to clean it up. 

What is most important though is that somewhere in the project you indicate that emn17 owns the copyright of some code that was licensed under the MIT license and is included in this project.

Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: Thanks @gyehuda. 
@diracdeltas, can you suggest which convention/approach you prefer?

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 3 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Got you. I merged over there. :+1:

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 4 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: agree. I switched to use `isIdentity()`.
And let's track the infinity point and `isInfinity()` that we're unsure at https://github.com/google/end-to-end/issues/340

File: src/javascript/crypto/e2e/vrf/vrf_test.html (hunk 7 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf_test.html]: completed in https://github.com/yahoo/coname/pull/25

File: src/javascript/crypto/e2e/vrf/vrf_test.html (hunk 8 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf_test.html]: addressed in https://github.com/yahoo/end-to-end/commit/5fb47531ee9ae5bbd44fc3ae745b20eed9443a8b#diff-565ba75ebd07abd46d0cb857452443b7

File: src/javascript/crypto/e2e/vrf/extra25519_test.html (hunk 11 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519_test.html:36]: tracking it at https://github.com/yahoo/end-to-end/issues/59

**Comment by adon-at-work** [comment]
After the last commit, the final issue to deal with in this PR: https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39409648

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  // hash := sha3.NewShake256()
+  // hash.Write(iiB[:]) // const length
+  // hash.Write(m)
+  // var hCheck [Size]byte
+  // hash.Read(hCheck[:])
+  // if !bytes.Equal(hCheck[:], vrf[:]) {
+  //  return false
+  // }
+  // hash.Reset()
+  hash = e2e.vrf.sha3Shake256(iiB.concat(m), 32);
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/vrf.js]: hash is undeclared

Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf.js]: gd catch

**Comment by diracdeltas** [comment]
Currently, the VRF module isn't included in any of the extension files, so the compiler isn't actually checking it for correctness. When I include the vrf module in the extension (such as by prepending `goog.require('e2e.vrf.verify');` to `launcher.js`), Closure compiler throws a lot of errors.

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+goog.provide('e2e.vrf.extra25519');
+
+goog.require('e2e.BigNum');
+goog.require('e2e.ecc.DomainParam');
+goog.require('e2e.ecc.Element');
+goog.require('e2e.ecc.PrimeCurve');
+goog.require('e2e.error.InvalidArgumentsError');
+
+var ed25519 = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.ED_25519);
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Declaring variables/functions like this will pollute the global namespace. Please make sure everything is contained in the module namespace. Ex: `e2e.vrf.extra255519.ed25519_ = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.ED_25519)`

The underscore at the end of a name indicates to the compiler that this variable is private (should not be accessed outside of the current module).

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+
+goog.provide('e2e.vrf.verify');
+
+goog.require('e2e.BigNum');
+goog.require('e2e.ecc.DomainParam');
+goog.require('e2e.ecc.PrimeCurve');
+goog.require('e2e.vrf.extra25519');
+goog.require('e2e.vrf.sha3Shake256');
+
+var PublicKeySize = 32,
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/vrf.js]: Should these be in all-caps? (for constants)

File: src/javascript/crypto/e2e/vrf/extra25519.js (hunk 13 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: makes sense.

File: src/javascript/crypto/e2e/vrf/vrf.js (hunk 14 above)
Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/vrf.js]: I can change them as private variables for this namespace with all-caps.

File: src/javascript/crypto/e2e/vrf/extra25519_test.html [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+<script>
+
+
+/**
+ * Tests that e2e.vrf.extra25519.fromBytesBaseGroup works, and no assertion error is thrown
+ */
+function testFromBytesBaseGroup() {
+	var pk = [194, 191, 96, 139, 106, 249, 24, 253, 198, 131, 88, 169, 100, 231, 7, 211, 70, 171, 171, 207, 24, 30, 150, 114, 77, 124, 240, 123, 191, 14, 29, 111];
+
+	e2e.vrf.extra25519.fromBytesBaseGroup(pk);
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/extra25519_test.html]: this test should probably contain some sort of assert statement.

Comment (adon-at-work) [inline src/javascript/crypto/e2e/vrf/extra25519_test.html]: gd catch. assert statements were once inlined in `fromBytesBaseGroup()`, but later removed. will update it accordingly

**Comment by adon-at-work** [comment]
> Currently, the VRF module isn't included in any of the extension files, so the compiler isn't actually checking it for correctness. When I include the vrf module in the extension (such as by prepending `goog.require('e2e.vrf.verify');` to `launcher.js`), Closure compiler throws a lot of errors.

the errors are cleared.

**Comment by adon-at-work** [comment]
I'm done with the suggested changes.

The build was failed due to an upstream commit (happened roughly an hour ago):
https://github.com/google/closure-templates/commit/e47f247bdfacf3c5a34015ced356e4fd8a0444a2

Tracking the issue at https://github.com/google/closure-templates/issues/57

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+ * Converts a uniform representative value to an element in Curve25519,
+ *  using the elligator map
+ * Returns an element that is either -A/(1+ur^2) or A - A/(1+ur^2),
+ *  where u = 2, A = 486662 (defined in curve25519)
+ *
+ * Refer to Section 5.2 in http://elligator.cr.yp.to/elligator-20130828.pdf
+ * @param {!e2e.ecc.Element} r the representative coordinate
+ * @return {!e2e.ecc.Element}
+ */
+function representativeToMontgomeryX(r) {
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: Please put this in the e2e.vrf.extra25519 namespace

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+
+
+/**
+ * Converts a Curve25519 element to an element in Ed25519
+ *
+ * Refer to p.8 in http://ed25519.cr.yp.to/ed25519-20110926.pdf
+ * @param {!e2e.ecc.Element} x the Curve25519 coordinate
+ * @return {!e2e.ecc.Element}
+ */
+function montgomeryXToEdwardsY(x) {
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/extra25519.js]: same as above

File: src/javascript/crypto/e2e/vrf/sha3shake256.js [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+
+/**
+ * @fileoverview Extends js-sha3 to support SHAKE256, and outputs byte array
+ */
+
+goog.provide('e2e.vrf.sha3Shake256');
+
+goog.require('goog.asserts');
+
+
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/sha3shake256.js]: all the variable declarations in the top-level namespace below should be in the e2e.vrf.sha3Shake256 namespace

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+goog.require('e2e');
+goog.require('e2e.BigNum');
+goog.require('e2e.ecc.DomainParam');
+goog.require('e2e.ecc.PrimeCurve');
+goog.require('e2e.vrf.extra25519');
+goog.require('e2e.vrf.sha3Shake256');
+
+
+/**
+ * @private Public Key Size
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/vrf.js:34]: nit: should be `PUBLIC_KEY_SIZE_`. Same for other constants below.

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+
+/**
+ * Converts an 32-byte hashed output to an Ed25519 point using Elligator
+ * This point is multiplied by 8.
+ *
+ * Refer to Section 5.2 in http://elligator.cr.yp.to/elligator-20130828.pdf
+ * @param {!e2e.ByteArray} m A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+function hashToCurve(m) {
```
Comment (diracdeltas) [inline src/javascript/crypto/e2e/vrf/vrf.js]: put in the e2e.vrf namespace

**Comment by adon-at-work** [comment]
@diracdeltas , all done. upgraded to use the latest js-sha3 too

**Comment by diracdeltas** [comment]
Thanks, LGTM! @andres-erbsen any objections to merging?

File: src/javascript/crypto/e2e/vrf/extra25519.js [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+
+
+/**
+ * Converts an 32-byte hashed output to an Ed25519 point using Elligator
+ *
+ * Refer to Section 5.2 in http://elligator.cr.yp.to/elligator-20130828.pdf
+ * @param {!e2e.ByteArray} h A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.extra25519.hashToEdwards = function(h) {
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/extra25519.js:159]: Please document that this is not constant time (because variable-length bignums aren't)

File: src/javascript/crypto/e2e/vrf/vrf.js [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+
+/**
+ * Converts an 32-byte hashed output to an Ed25519 point using Elligator
+ * This point is multiplied by 8.
+ *
+ * Refer to Section 5.2 in http://elligator.cr.yp.to/elligator-20130828.pdf
+ * @param {!e2e.ByteArray} m A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.hashToCurve = function(m) {
```
Comment (andres-erbsen) [inline src/javascript/crypto/e2e/vrf/vrf.js:180]: Please document that this is not constant time and thus should not be used with secret inputs. reason: hashToCurve isn't.

**Comment by andres-erbsen** [comment]
No objections. I did add two one-line documentation nits about functions that are not constant-time-ish, though.
